receiver/kafkareceiver/consumer_sarama.go

```diff
@@ -105,42 +96,88 @@ func (c *saramaConsumer) Start(_ context.Context, host component.Host) error {
 
 	c.consumeLoopClosed = make(chan struct{})
 	c.started = true
-	go c.consumeLoop(handler)
+	c.closing = make(chan struct{})
+	go c.consumeLoop(handler, host)
 	return nil
 }
 
-func (c *saramaConsumer) consumeLoop(handler sarama.ConsumerGroupHandler) {
+func (c *saramaConsumer) consumeLoop(handler sarama.ConsumerGroupHandler, host component.Host) {
 	defer close(c.consumeLoopClosed)
+	defer componentstatus.ReportStatus(host, componentstatus.NewEvent(componentstatus.StatusStopped))
+	componentstatus.ReportStatus(host, componentstatus.NewEvent(componentstatus.StatusStarting))
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		select {
+		case <-ctx.Done():
+		case <-c.closing:
+			componentstatus.ReportStatus(host, componentstatus.NewEvent(componentstatus.StatusStopping))
+			cancel()
+		}
+	}()
+
+	// kafka.NewSaramaConsumerGroup (actually sarama.NewConsumerGroup)
+	// may perform synchronous operations that can fail due to transient
+	// errors, so we retry until it succeeds or the context is canceled.
+	var consumerGroup sarama.ConsumerGroup
+	err := backoff.Retry(func() (err error) {
+		consumerGroup, err = kafka.NewSaramaConsumerGroup(ctx, c.config.ClientConfig, c.config.ConsumerConfig)
+		if err != nil {
+			if ctx.Err() == nil {
+				// We only report an error if the context is not canceled.
+				// If the context is canceled it means the receiver is
+				// shutting down, which will lead to reporting StatusStopped
+				// when consumeLoop exits.
+				c.settings.Logger.Error("Error creating consumer group", zap.Error(err))
+				componentstatus.ReportStatus(host, componentstatus.NewRecoverableErrorEvent(err))
+			}
+			return err
+		}
+		return nil
+	}, backoff.WithContext(
+		backoff.NewExponentialBackOff(backoff.WithMaxElapsedTime(0)),
```

Comment:
Should this value be zero for passing with MaxElaspsedTime??

Reply:
I've added a comment to clarify that